libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -854,6 +887,60 @@ class MapboxNavigation(
         }
     }
 
+    private fun createInternalFallbackVersionsObserver() = object : FallbackVersionsObserver {
+        override fun onFallbackVersionsFound(versions: List<String>) {
+            if (versions.isNotEmpty()) {
+                // the last version in the list is the latest one
+                val tilesVersion = versions.last()
+                recreateNavigatorInstance(isFallback = true, tilesVersion = tilesVersion)
+                navigationVersionSwitchObservers.forEach {
+                    it.onSwitchToFallbackVersion(tilesVersion)
+                }
+            } else {
+                logger.d(
+                    TAG,
+                    Message(
+                        "FallbackVersionsObserver.onFallbackVersionsFound called with an empty " +
+                            "versions list, navigator can't be recreated."
+                    )
+                )
+            }
+        }
+
+        override fun onCanReturnToLatest() {
+            recreateNavigatorInstance(
+                isFallback = false,
+                tilesVersion = navigationOptions.routingTilesOptions.tilesVersion
+            )
+            navigationVersionSwitchObservers.forEach {
+                it.onSwitchToTargetVersion()
+            }
+        }
+    }
+
+    private fun recreateNavigatorInstance(isFallback: Boolean, tilesVersion: String) {
+        logger.d(
+            TAG,
+            Message(
+                "recreateNavigatorInstance(). " +
+                    "isFallback = $isFallback, tilesVersion = $tilesVersion"
+            )
+        )
+        val route = tripSession.route
+        val routeProgress = tripSession.getRouteProgress()
+        navigator.recreate(
```

Comment:
Is this operation something costly / blocking? cc @SiarheiFedartsou @mskurydin @DmitryAk

Reply:
can be potentially heavy (involves a bunch of things with multi-threading).